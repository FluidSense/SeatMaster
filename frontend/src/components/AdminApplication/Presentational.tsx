import React from 'react';
import { IApplicationInfoObject } from '../ApplicationReview';
import ApplicationOverview from '../ApplicationReview/ApplicationOverview';

interface IProps {
  application: IApplicationInfoObject;
}

const Presentational: React.FunctionComponent<IProps> = (props) => {
  const { application } = props;

  return (
    <div className="main-content">
      <ApplicationOverview applicationInfo={application} title={application.fullname}/>
    </div>
  );
};

export default Presentational;
