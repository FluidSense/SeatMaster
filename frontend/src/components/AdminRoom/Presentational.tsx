import AlertStripe from 'nav-frontend-alertstriper';
import HovedKnapp from 'nav-frontend-knapper';
import { Input, Textarea } from 'nav-frontend-skjema';
import { Sidetittel } from 'nav-frontend-typografi';
import React, { ChangeEvent, SyntheticEvent } from 'react';
import { ETIKETT_WARNING } from '../commonConstants';
import {
  _ALERT_CREATED_MESSAGE,
  _BUTTON_CREATE_ROOM,
  _BUTTON_DELETE_ROOM,
  _BUTTON_UPDATE_ROOM,
  _INPUT_LABEL_NAME,
  _INPUT_LABEL_NOTES,
  _TITLE_CREATE_NEW_ROOM,
  _TITLE_UPDATE_NEW_ROOM,
} from './strings';

interface IProps {
  roomName: string;
  roomNotes: string;
  setNotes: (roomNotes: ChangeEvent<HTMLInputElement>) => void;
  setName: (roomName: ChangeEvent<HTMLInputElement>) => void;
  buttonDisabled: boolean;
  onClick: () => void;
  deleteRoom: () => void;
  showAlert: boolean;
  alertMessage?: string;
  roomExists: boolean;
}

const Presentational: React.FunctionComponent<IProps> = (props) => {
  const {
    roomName,
    roomNotes,
    roomExists,
    setNotes,
    setName,
    buttonDisabled,
    onClick,
    deleteRoom,
    showAlert,
    alertMessage,
  } = props;
  const displayAlert =
    showAlert
      ? (
        <AlertStripe solid={true} type={ETIKETT_WARNING}>
          {alertMessage}
        </AlertStripe>)
      : null;
  const titleText = roomExists ? _TITLE_UPDATE_NEW_ROOM : _TITLE_CREATE_NEW_ROOM;
  const buttonText = roomExists ? _BUTTON_UPDATE_ROOM : _BUTTON_CREATE_ROOM;
  const deleteButton =
    roomExists
      ? (
        <HovedKnapp id={'delete-room-button'} type="fare" onClick={deleteRoom}>
          {_BUTTON_DELETE_ROOM}
        </HovedKnapp>)
      : null;
  // TextArea returns the wrong type, so its type has to be forced
  const assertEventType = (event: SyntheticEvent<EventTarget, Event>) => {
    const changeEvent = event as ChangeEvent<HTMLInputElement>;
    setNotes(changeEvent);
  };

  return (
    <>
      <Sidetittel>{titleText}</Sidetittel>
      {displayAlert}
      <Input
        id={'input-room-name'}
        onChange={setName}
        value={roomName}
        label={_INPUT_LABEL_NAME}
        bredde={'XL'}
      />
      <Textarea
        id={'input-room-notes'}
        onChange={assertEventType}
        value={roomNotes}
        label={_INPUT_LABEL_NOTES}
      />
      <div id="state-buttons">
        <HovedKnapp
          id={'create-room-button'}
          type="hoved"
          disabled={buttonDisabled}
          onClick={onClick}
        >
          {buttonText}
        </HovedKnapp>
        {deleteButton}
      </div>
    </>
  );
};

export default Presentational;
