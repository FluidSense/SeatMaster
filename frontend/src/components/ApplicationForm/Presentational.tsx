import AlertStripe from 'nav-frontend-alertstriper';
import KnappBase from 'nav-frontend-knapper';
import * as React from 'react';
import { postApplicationForm } from '../../API/calls';
import { IRegisteredUserState } from '../RegisterUser/reducer';
import { IRoom } from '../ViewRooms';
import ApplicationFormComments from './ApplicationFormComments';
import ApplicationFormPersonal from './ApplicationFormPersonal';
import ApplicationFormPreferences from './ApplicationFormPreferences';

import { _ALERT_USER_ERROR } from './Strings';

interface IProps {
  userInformation: IRegisteredUserState;
  rooms: IRoom[];
  changeModal: (modalOpen: boolean) => void;
  getRooms: () => void;
}

interface IState {
  room?: IRoom;
  partner: boolean;
  partnerUsername: string;
  needs: boolean;
  needsText: string;
  infoText: string;
  keepSeat: boolean;
  loading: boolean;
  error: string;
  [key: string]: string | boolean | IRoom | undefined;
}

export class Presentational extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      error: '',
      infoText: '',
      keepSeat: false,
      loading: false,
      needs: false,
      needsText: '',
      partner: false,
      partnerUsername: '',
      room: undefined,
    };
  }

  public componentDidMount() {
    this.props.getRooms();
  }

  public render() {
    const { userInformation } = this.props;
    const alertBox = this.state.error ? this.alertUser(_ALERT_USER_ERROR) : undefined;
    return (
      <>
        {alertBox}
        <form
          onSubmit={this.onSubmitForm}
          id="new-application-form"
        >
          <ApplicationFormPersonal
            fullname={userInformation.fullname}
            email={userInformation.email}
            status={userInformation.masterStatus}
          />
          <ApplicationFormPreferences
            updateApplicationFormData={this.updateApplicationFormData}
            partner={this.state.partner}
            rooms={this.props.rooms}
          />
          <ApplicationFormComments
            updateApplicationFormData={this.updateApplicationFormData}
            needs={this.state.needs}
          />
          <KnappBase
            id="submit-application"
            type="hoved"
            htmlType="submit"
            autoDisableVedSpinner={true}
            spinner={this.state.loading}
          >
            Submit
          </KnappBase>
        </form>
      </>
    );
  }

  private updateApplicationFormData = (item: React.FormEvent) => {
    const eventTarget = item.target as HTMLFormElement;
    const name: string = eventTarget.name;
    const value: string | boolean =
      eventTarget.type === 'checkbox' ? eventTarget.checked : eventTarget.value;
    this.setState({ [name]: value });
  }

  private onSubmitForm = (e: React.FormEvent) => {
    e.preventDefault();
    this.setState({ loading: true });

    postApplicationForm({
      comments: this.state.infoText,
      needs: this.state.needs ? this.state.needsText : '',
      partnerUsername: this.state.partnerUsername,
      preferredRoom: this.state.room ? this.state.room.name : this.state.room,
      seatRollover: this.state.keepSeat,
    })
      .then(
        // On fullfilled promise:
        () => {
          this.props.changeModal(true);
          this.setState({ loading: false, error: '' });
        },
        // On rejected promise:
        () => {
          this.setState({ loading: false, error: _ALERT_USER_ERROR });
        });
  }

  private alertUser = (text: string) => (
    <AlertStripe
      type="advarsel"
      solid={true}
    >
      {text}
    </AlertStripe>
  )
}

export default Presentational;
