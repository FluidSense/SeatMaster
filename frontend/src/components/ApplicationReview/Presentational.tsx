import KnappBase from 'nav-frontend-knapper';
import React from 'react';
import { Link } from 'react-router-dom';
import { APP_NOT_FOUND } from '../Login/constants';
import { IApplicationInfoObject } from './index';
import InfoPanel from './InfoPanel';
import InformationList from './InformationList';
import {
  _EDIT_APPLICATION,
  _EMAIL,
  _MASTER_STATUS,
  _NAME,
  _NEEDS,
  _NEW_APPLICATION,
  _PARTNER,
  _PHONE,
  _PREFERRED_ROOM,
  _REVIEW_APPLICATION,
  _SEAT_ROLLOVER,
} from './strings';
import { Sidetittel } from 'nav-frontend-typografi';

interface IProps {
  applicationInfo: IApplicationInfoObject;
}

export interface IUserInfoObject {
  [_NAME]?: string;
  [_EMAIL]?: string;
  [_PHONE]?: string;
  [_MASTER_STATUS]?: string;
}

export interface IRoomInfObject {
  [_PARTNER]?: string;
  [_PREFERRED_ROOM]?: string;
  [_SEAT_ROLLOVER]?: string;
}

const Presentational: React.FunctionComponent<IProps> = (props) => {
  const { applicationInfo } = props;
  const userInfoObject: IUserInfoObject = {
    [_NAME]: applicationInfo.user !== undefined
      ? applicationInfo.user.username
      : applicationInfo.fullname,
    [_EMAIL]: applicationInfo.email,
    [_PHONE]: applicationInfo.phone,
    [_MASTER_STATUS]: applicationInfo.status,
  };
  const roomInfoObject = {
    [_PARTNER]: applicationInfo.partner,
    [_PREFERRED_ROOM]: applicationInfo.room,
    [_SEAT_ROLLOVER]: applicationInfo.seatRollover,
  };
  if (applicationInfo.applicationStatus === APP_NOT_FOUND) {
    return (
      <Link to="/application">
        <KnappBase type="hoved" htmlType="button" id="newAppButton">{_NEW_APPLICATION}</KnappBase>
      </Link>
    );
  }
  return (
    <>
      <Sidetittel>{_REVIEW_APPLICATION}</Sidetittel>
      <div id="user-information">{<InformationList information={userInfoObject} />}</div>
      <div id="room-information">{<InformationList information={roomInfoObject} />}</div>
      <div className={'needs-information'}>
        {<InfoPanel title={_NEEDS} text={applicationInfo.needs} />}
      </div>
      <KnappBase id="edit-application" type="hoved">{_EDIT_APPLICATION}</KnappBase>
    </>
  );
};

export default Presentational;
