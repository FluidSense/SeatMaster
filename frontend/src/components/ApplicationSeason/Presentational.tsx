import EtikettBase from 'nav-frontend-etiketter';
import React from 'react';
import { ETIKETT_INFO } from '../commonConstants';
import { IApplicationSeason } from './reducer';
import {
  _SEASON_CLOSED,
  _SEASON_END,
  _SEASON_OPEN,
  _SEASON_START,
} from './strings';

interface IApplicationSeasonProps {
  applicationSeason: IApplicationSeason;
  currentTime: Date;
}

type validPreText = typeof _SEASON_START | typeof _SEASON_END;

const formatSeasonString = (preText: validPreText, time: Date) => {
  const day = time.getDate();
  const month = time.getMonth() + 1;
  const year = time.getFullYear();
  return `${preText} - ${day}/${month}/${year}`;
};

const Presentational: React.FunctionComponent<IApplicationSeasonProps> = (props) => {
  if (props.applicationSeason === undefined) return null;
  const { applicationPeriodEnd, applicationPeriodStart } = props.applicationSeason;
  const { currentTime } = props;

  let t: string = '';
  if (currentTime < applicationPeriodStart) {
    t = formatSeasonString(_SEASON_START, applicationPeriodStart);
  } else {
    t = _SEASON_OPEN;
  }
  const seasonStartText: string = t;

  let p: any = '';
  if (currentTime < applicationPeriodEnd) {
    p = formatSeasonString(_SEASON_END, applicationPeriodEnd);
  } else {
    p = _SEASON_CLOSED;
  }
  const seasonEndText: string = p;

  return (
    <>
      <EtikettBase type={ETIKETT_INFO}>{seasonStartText}</EtikettBase>
      <EtikettBase type={ETIKETT_INFO}>{seasonEndText}</EtikettBase>
    </>
  );
};

export default Presentational;
