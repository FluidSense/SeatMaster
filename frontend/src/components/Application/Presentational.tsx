import * as React from 'react';
import ApplicationForm from '../ApplicationForm';
import ConfirmationModal from './../ApplicationForm/ConfirmationModal';

interface IProps {
  modalIsOpen: boolean;
  changeModal: (modalOpen: boolean) => void;
}

export const Presentational: React.FunctionComponent<IProps> = (props) => {
  return (
    <>
      <ConfirmationModal modalIsOpen={props.modalIsOpen} changeModal={props.changeModal} />
      <ApplicationForm changeModal={props.changeModal} />
    </>
  );
};

export default Presentational;
