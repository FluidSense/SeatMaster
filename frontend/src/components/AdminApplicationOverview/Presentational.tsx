import { Sidetittel } from 'nav-frontend-typografi';
import React, { FunctionComponent } from 'react';
import { IApplication } from '../Application';
import { TitleAndSpinner } from '../LoadingPageSpinner/TitleAndSpinner';
import SearchBar, { searchBarEvent } from '../SearchBar';
import { IRoom } from '../ViewRooms';
import ApplicationLink from './ApplicationLink';
import ApproveSeated from './ApproveSeated';
import { FETCHING_APPLICATION_DATA } from './constants';
import PutOnWaitingList from './PutOnWaitingList';
import { _APPLICATIONS_TITLE } from './strings';

interface IProps {
  applications: IApplication[];
  rooms: IRoom[];
  fetching: string;
  approve: (ids: number[]) => void;
  filterFunction: (event: searchBarEvent) => void;
  putWaiting: (ids: number[]) => void;
}
const title = <Sidetittel>{_APPLICATIONS_TITLE}</Sidetittel>;
const emptyPage = (filterFunction: (event: searchBarEvent) => void) => (
  <div className="main-content">
    {title}
    <SearchBar filterFunction={filterFunction}/>
</div>);
const emptyPageWithLoader = (<TitleAndSpinner title={_APPLICATIONS_TITLE}/>);

const Presentational: FunctionComponent<IProps> = (props) => {
  const { applications, rooms, fetching, filterFunction, approve, putWaiting } = props;
  if (fetching === FETCHING_APPLICATION_DATA) return emptyPageWithLoader;
  if (!applications || !applications.length) return emptyPage(filterFunction);
  const applicationList = applications.map(application => (
    <ApplicationLink key={application.id} application={application} rooms={rooms} />
    ));
  return (
    <div className="main-content">
      <div className="title-and-button">
        <Sidetittel>
          {_APPLICATIONS_TITLE}
        </Sidetittel>
        <PutOnWaitingList putWaiting={putWaiting} applications={applications}/>
        <ApproveSeated approve={approve} applications={applications}/>
      </div>
      <SearchBar filterFunction={filterFunction}/>
      {applicationList}
    </div>
  );
};

export default Presentational;
