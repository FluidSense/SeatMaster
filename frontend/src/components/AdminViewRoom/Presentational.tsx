import KnappBase from 'nav-frontend-knapper';
import { Sidetittel } from 'nav-frontend-typografi';
import React from 'react';
import { Redirect } from 'react-router';
import { IPostAdminApplicationForm, IUser } from '../../API/interfaces';
import CSVButton from '../AdminRoom/CSVButton';
import { IApplication } from '../Application';
import { ISeat } from '../Seats';
import { IRoom } from '../ViewRooms';
import SeatPlacer from './SeatPlacer';
import { _EDIT_ROOM_BUTTON } from './strings';

interface IStateProps {
  match: {
    params: {
      id: string,
    },
  };
  applications?: IApplication[];
  rooms: IRoom[];
}

interface IDispatchProps {
  assign: (user: IUser, seat: ISeat) => void;
  delete: (seat: ISeat) => void;
  getAllApplications: () => void;
  fetchRooms: () => void;
  fetchRoom: (roomId: number) => void;
  updateApplication: (id: number, app: IPostAdminApplicationForm) => void;

}

interface IState {
  redirect: boolean;
  fetchedRooms: boolean;
  fetchedApplications: boolean;
}

type Props = IStateProps & IDispatchProps;

class Presentational extends React.Component<Props, IState> {
  constructor(props: Props) {
    super(props);
    this.state = {
      fetchedApplications: false,
      fetchedRooms: false,
      redirect: false,
    };
  }

  public componentDidMount() {
    if (!this.props.rooms.length && !this.state.fetchedRooms) {
      this.setState({ fetchedRooms: true });
      this.props.fetchRooms();
    }
  }

  public render() {
    const { applications, rooms, fetchRoom } = this.props;
    const roomId = parseInt(this.props.match.params.id, 10);
    const { redirect } = this.state;
    const room = rooms.find(iterroom => iterroom.id === roomId);
    if (!room) return null; // Insert spinner here
    const seats = this.seatsPlacers(room.seats.seats, applications);
    if (redirect) {
      return (
        <Redirect
          to={{ room, pathname: `/admin/rooms/${room.id}/update-room` }}
        />);
    }

    return (
      <div className="main-content">
        <div className="title-and-button">
          <Sidetittel>{room.name}</Sidetittel>
          <CSVButton room={room} fetchRoomInfo={fetchRoom} />
          <KnappBase type="hoved" onClick={this.setRedirect}>
            {_EDIT_ROOM_BUTTON}
          </KnappBase>
        </div>
        <h2>Info</h2>
        <p>{room.info}</p>
        <div>
          {seats}
        </div>
      </div>
    );
  }

  private setRedirect = () => {
    this.setState({ redirect: true });
  }

  private seatsPlacers = (seats: ISeat[], applications?: IApplication[]) => {
    if (!(applications && applications.length) && !this.state.fetchedApplications) {
      this.setState({ fetchedApplications: true });
      this.props.getAllApplications();
      return;
    }
    const apps = applications ? applications : [];
    const users = apps.map(application => application.user);
    return seats.map((seat, index) => {
      return (
        <SeatPlacer
          delete={this.props.delete}
          key={index}
          seat={seat}
          users={users}
          assign={this.props.assign}
          applications={this.props.applications}
          updateApplication={this.props.updateApplication}
        />
      );
    });
  }
}

export default Presentational;
