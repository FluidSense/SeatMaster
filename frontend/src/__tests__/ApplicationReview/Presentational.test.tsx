import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import Presentational from '../../components/ApplicationReview/Presentational';

const applicationInfoMock = {
  applicationStatus: 'SUBMITTED',
  email: 'test@test.com',
  fullname: 'Teser Testersen',
  partner: 'None',
  phone: '91100999',
  room: 'Space Commander',
  seatRollover: 'Yes',
  status: 'Master of masters',
};

describe('Application review presentational', () => {
  it('Renders correctly', () => {
    const wrapper = shallow(<Presentational applicationInfo={applicationInfoMock} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
