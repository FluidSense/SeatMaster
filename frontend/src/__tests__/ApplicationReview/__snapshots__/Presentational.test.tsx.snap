// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application review presentational Renders correctly 1`] = `
<ApplicationOverview
  applicationInfo={
    Object {
      "applicationStatus": "SUBMITTED",
      "email": "test@test.com",
      "fullname": "Teser Testersen",
      "partner": "None",
      "phone": "91100999",
      "room": "Space Commander",
      "seatRollover": "Yes",
      "status": "Master of masters",
    }
  }
/>
`;
